### ISSUE TITLE
Conch breaks __builtin__ pow with floats

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4803]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| spiv reported|
|-|-|
|Trac ID|trac#4803|
|Type|defect|
|Created|2011-01-14 16:20:44Z|
|Branch|https://github.com/twisted/twisted/tree/pow-floats-4803|

```
>>> pow(2.5, 2)
6.25
>>> import twisted.conch.ssh.common
>>> pow(2.5, 2)
mpz(4)
>>> 
```

I'll make a fix for this.

<details><summary>Searchable metadata</summary>

```
trac-id__4803 4803
type__defect defect
reporter__spiv spiv
priority__normal normal
milestone__ 
branch__branches_pow_floats_4803 branches/pow-floats-4803
branch_author__spiv spiv
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1295022044000000 1295022044000000
changetime__1295047248000000 1295047248000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv set owner to spiv<br>spiv set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|spiv removed owner<br>spiv set status to `new`|
|-|-|

I have a fix in bzr at lp:~spiv/twisted/pow-floats-4803.  You can see the diff here: <http://bazaar.launchpad.net/~spiv/twisted/pow-floats-4803/revision/15708>.

This version is very slightly slower due to the type-checking, but still much faster than the builtin pow for sufficiently large integers.  And correctness > speed anyway.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_3"></a>|@jml set owner to spiv|
|-|-|

Thanks for the patch. It looks good to me, on the whole.

We discussed in person whether the type checking could be replaced with a check like `if x == int(x)`, and decided that it would be too inefficient in the cases where `x` is already an `mpz`.

There are a few places in the tests where you've neglected to add a space after a comma. Please fix those before landing.

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|spiv commented|
|-|-|

(In [30447]) Branch to pow-floats-4803

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|spiv commented|
|-|-|

(In [30448]) Don't break pow(base, exp) where base is not an integer.  Refs [#4803](https://github.com/twisted/twisted/issues/4803).

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|spiv set status to `closed`|
|-|-|

(In [30456]) Merge pow-floats-4803: Don't break builtin pow with non-integer bases.

Author: spiv
Reviewers: jml, exarkun
Fixes: [#4803](https://github.com/twisted/twisted/issues/4803)

Fix the _fastpow hack in twisted.ssh.common so that non-integer values
passed to pow aren't silently converted to integers.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
